### PR TITLE
Feature/refactor sm

### PR DESCRIPTION
**Pull request by JorxSS**
Refactor of the Script Manager:
- Now we use a library for hot reloading. 
- Encapsulated code for DLL.
- Moved Scripts and their solution to assets folder.
- Now the scripts we create will be added automatically to the Scripting project.
- Now we don't need to re-compile every time we make a modification in a Script, if you just modify an Script and save it, when you go back to the engine, the engine will compile the scripts for you.

### CODE REVIEW

File: Engine/DLL/DLLManager.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+void DLLManager::CheckGameplayFolderStatus()
+{
+	last_timestamp_script_folder = scripts_folder->GetModificationTimestamp();
+	if (last_timestamp_script_folder != init_timestamp_script_folder || CheckFolderTimestamps())
+	{
+		std::thread(&DLLManager::CheckCompilation, this).detach();
+	}
+}
```
Comment (Dartemiss) [inline Engine/DLL/DLLManager.cpp:96]: Just a question, why are u calling CheckCompilation like &DLLManager::CheckCompilation on the thread and not CheckCompilation as is defined on the same class.

File: Engine/DLL/DLLManager.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+void DLLManager::CompileGameplayProject() const
+{
+	APP_LOG_INFO("Change detected in the Gameplay System, compilation in process.");
+	std::wstring msbuild_path(MSBUILD_PATH);
+	std::string msbuild_path_to_string(msbuild_path.begin(), msbuild_path.end());
+	std::string command('\"' + msbuild_path_to_string + COMMAND_FOR_COMPILING);
```
Comment (Dartemiss) [inline Engine/DLL/DLLManager.cpp:103]: Didn't know about wstring, why do you use it only for msbuild_path and then use normal string?

File: Engine/DLL/DLLManager.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	system(command.c_str());
+
+}
+
+void DLLManager::CheckCompilation()
+{
+	if (!compiling) 
+	{
+		compiling = true;
+		std::thread compilator = std::thread(&DLLManager::CompileGameplayProject, this);
```
Comment (Dartemiss) [inline Engine/DLL/DLLManager.cpp:113]: Same question as before

File: Engine/Helper/Utils.cpp [hunk 4]
```diff
@@ -92,33 +92,11 @@ std::vector<float> Utils::GetVertices(const AABB2D& box)
...
-		return true;
-	}
-	else
-	{
-		return false;
-	}
-
+	char current_path[MAX_PATH];
+	GetCurrentDirectory(MAX_PATH, current_path);
+	path = current_path;
```
Comment (Dartemiss) [inline Engine/Helper/Utils.cpp:99]: This shouldnt be const?

**Review by Dartemiss — APPROVED** [review]
Amazing job! Now our gameplay systems have reached lvl super sayan 2 xD

File: Engine/DLL/DLLManager.cpp (hunk 2 above)
Comment (JorxSS) [reply Engine/DLL/DLLManager.cpp:103]: MSBUILD_PATH's value type is given by Visual Studio variable. So to convert it to string in order to use it the method to cast that into a string is using wstring

File: Engine/DLL/DLLManager.cpp (hunk 1 above)
Comment (JorxSS) [reply Engine/DLL/DLLManager.cpp:96]: If I'm not mistaken, std::thread in order to call a function that is member of a class, need to pass this function as it is there, creating a callback function. But to call a member function, we need a object and thats why also im passing "this".

File: Engine/Helper/Utils.cpp (hunk 4 above)
Comment (JorxSS) [reply Engine/Helper/Utils.cpp:99]: I think that cannot be const because is an static void function.

Comment (Dartemiss) [reply Engine/Helper/Utils.cpp:99]: Yes my bad didn't see it was static

**Review by marcorod94 — APPROVED** [review]
Really nice job! looks great

**Pull request merged**